app/src/main/java/ru/yandex/architectureproject/presentation/viewmodel/TaskViewModel.kt

```diff
@@ -0,0 +1,69 @@
+package ru.yandex.architectureproject.presentation.viewmodel
+
+import androidx.lifecycle.ViewModel
+import androidx.lifecycle.viewModelScope
+import kotlinx.coroutines.CoroutineDispatcher
+import kotlinx.coroutines.Job
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.flow.StateFlow
+import kotlinx.coroutines.flow.asStateFlow
+import kotlinx.coroutines.flow.catch
+import kotlinx.coroutines.flow.distinctUntilChanged
+import kotlinx.coroutines.flow.onStart
+import kotlinx.coroutines.job
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.withContext
+import ru.yandex.architectureproject.domain.AddTaskUseCase
+import ru.yandex.architectureproject.domain.CompleteTaskUseCase
+import ru.yandex.architectureproject.domain.DeleteTaskUseCase
+import ru.yandex.architectureproject.domain.GetAllTasksUseCase
+import ru.yandex.architectureproject.domain.IncompleteTaskUseCase
+import ru.yandex.architectureproject.presentation.state.TaskAction
+import ru.yandex.architectureproject.presentation.state.TaskState
+
+class TaskViewModel(
+    private val addTaskUseCase: AddTaskUseCase,
+    private val deleteTaskUseCase: DeleteTaskUseCase,
+    private val getAllTasksUseCase: GetAllTasksUseCase,
+    private val completeTaskUseCase: CompleteTaskUseCase,
+    private val incompleteTaskUseCase: IncompleteTaskUseCase,
+    private val ioDispatcher: CoroutineDispatcher,
+) : ViewModel() {
+    private val _state = MutableStateFlow<TaskState>(TaskState.Loading)
+    val state: StateFlow<TaskState> = _state.asStateFlow()
+
+    private val taskForDeletionJobMap: MutableMap<Int, Job?> = mutableMapOf()
+
+    init {
+        reduce(TaskAction.LoadTasks)
+    }
+
+    fun reduce(action: TaskAction) = viewModelScope.launch {
+        when (action) {
+            is TaskAction.LoadTasks -> loadTasks()
+            is TaskAction.AddTask -> addTaskUseCase(action.task)
+
+            is TaskAction.UpdateTaskStatus -> {
+                if (action.done) {
+                    taskForDeletionJobMap[action.taskId] = this.coroutineContext.job
+                    completeTaskUseCase(action.taskId)
+                } else {
+                    taskForDeletionJobMap[action.taskId]?.cancel()
```

Comment:
По-хорошему тут ещё лучше удалить джобу из мапы:
```suggestion
                    taskForDeletionJobMap[action.taskId]?.remove()?.cancel()
```